File: src/game/cards.ts
import type { Locale } from "./i18n.ts";
import { t } from "./i18n.ts";
import type { Card, Rank, Suit } from "./types.ts";

export const SUITS: Suit[] = ["clubs", "diamonds", "spades", "hearts"];
export const RANKS: Rank[] = [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14];

export const SUIT_ORDER: Record<Suit, number> = {
  clubs: 0,
  diamonds: 1,
  spades: 2,
  hearts: 3,
};

const SUIT_CHAR: Record<Suit, string> = {
  clubs: "C",
  diamonds: "D",
  spades: "S",
  hearts: "H",
};

const CHAR_SUIT: Record<string, Suit> = {
  C: "clubs",
  D: "diamonds",
  S: "spades",
  H: "hearts",
};

export function rankCode(rank: Rank): string {
  if (rank === 11) return "J";
  if (rank === 12) return "Q";
  if (rank === 13) return "K";
  if (rank === 14) return "A";
  return String(rank);
}

export function cardId(suit: Suit, rank: Rank): string {
  return `${SUIT_CHAR[suit]}${rankCode(rank)}`;
}

export function parseCard(id: string): Card {
  const suit = CHAR_SUIT[id[0] ?? ""];
  const rankStr = id.slice(1);
  const named: Record<string, Rank> = { J: 11, Q: 12, K: 13, A: 14 };
  const rank = named[rankStr] ?? (Number(rankStr) as Rank);
  if (!suit || !rank) throw new Error(`Invalid card: ${id}`);
  return { id, suit, rank };
}

export function makeCard(suit: Suit, rank: Rank): Card {
  return { id: cardId(suit, rank), suit, rank };
}

export function makeDeck(): Card[] {
  const deck: Card[] = [];
  for (const suit of SUITS) {
    for (const rank of RANKS) {
      deck.push(makeCard(suit, rank));
    }
  }
  return deck;
}

export function createRng(seed?: number): () => number {
  if (seed == null) return Math.random;
  let s = seed >>> 0;
  return () => {
    s = (Math.imul(s, 1664525) + 1013904223) >>> 0;
    return s / 4294967296;
  };
}

export function shuffle<T>(items: T[], rng: () => number = Math.random): T[] {
  const next = items.slice();
  for (let i = next.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    const a = next[i]!;
    next[i] = next[j]!;
    next[j] = a;
  }
  return next;
}

export function sortHand(hand: Card[]): Card[] {
  return hand.slice().sort((a, b) => {
    const suitDiff = SUIT_ORDER[a.suit] - SUIT_ORDER[b.suit];
    if (suitDiff !== 0) return suitDiff;
    return a.rank - b.rank;
  });
}

export function isQueenOfSpades(card: Card): boolean {
  return card.id === "SQ";
}

export function isJackOfDiamonds(card: Card): boolean {
  return card.id === "DJ";
}

export function isTwoOfClubs(card: Card): boolean {
  return card.id === "C2";
}

export function isHeart(card: Card): boolean {
  return card.suit === "hearts";
}

export function isPenaltyCard(card: Card): boolean {
  return isHeart(card) || isQueenOfSpades(card);
}

export function cardLabel(card: Card, locale: Locale = "en"): string {
  const copy = t(locale);
  if (isQueenOfSpades(card)) return copy.queenOfSpades;
  if (isJackOfDiamonds(card)) return copy.jackOfDiamonds;
  const rank =
    card.rank === 14
      ? copy.rank.ace
      : card.rank === 13
        ? copy.rank.king
        : card.rank === 12
          ? copy.rank.queen
          : card.rank === 11
            ? copy.rank.jack
            : String(card.rank);
  return `${copy.suitCap[card.suit]} ${rank}`;
}

export const cardLabelNb = (card: Card) => cardLabel(card, "nb");

export function sameCard(a: Card, b: Card): boolean {
  return a.id === b.id;
}

export function findCard(hand: Card[], id: string): Card | undefined {
  return hand.find((c) => c.id === id);
}

export function withoutCards(hand: Card[], ids: string[]): Card[] {
  const drop = new Set(ids);
  return hand.filter((c) => !drop.has(c.id));
}

File: native/src/cards.rs
use serde::{Deserialize, Serialize};

pub type Suit = u8; // 0 clubs, 1 diamonds, 2 spades, 3 hearts
pub type Rank = u8; // 2..=14
pub type PlayerId = usize; // 0..3

pub const CLUBS: Suit = 0;
pub const DIAMONDS: Suit = 1;
pub const SPADES: Suit = 2;
pub const HEARTS: Suit = 3;

pub const SUITS: [Suit; 4] = [CLUBS, DIAMONDS, SPADES, HEARTS];
pub const RANKS: [Rank; 13] = [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14];
pub const SUIT_ORDER: [u8; 4] = [0, 1, 2, 3];

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Card {
    pub id: u8, // packed: suit * 16 + rank
}

impl Card {
    pub fn new(suit: Suit, rank: Rank) -> Self {
        Self { id: suit * 16 + rank }
    }
    pub fn suit(self) -> Suit {
        self.id / 16
    }
    pub fn rank(self) -> Rank {
        self.id % 16
    }
}

pub fn rank_code(rank: Rank) -> &'static str {
    match rank {
        11 => "J",
        12 => "Q",
        13 => "K",
        14 => "A",
        r => match r {
            2 => "2",
            3 => "3",
            4 => "4",
            5 => "5",
            6 => "6",
            7 => "7",
            8 => "8",
            9 => "9",
            10 => "10",
            _ => "",
        },
    }
}

pub fn make_deck() -> Vec<Card> {
    let mut deck = Vec::with_capacity(52);
    for &suit in &SUITS {
        for &rank in &RANKS {
            deck.push(Card::new(suit, rank));
        }
    }
    deck
}

pub struct Rng {
    state: u32,
}

impl Default for Rng {
    fn default() -> Self {
        Self::new()
    }
}

impl Rng {
    pub fn new() -> Self {
        let seed = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_nanos() as u32)
            .unwrap_or(0xdead_beef)
            .wrapping_mul(2654435761);
        Self::from_seed(seed | 1)
    }
    pub fn from_seed(seed: u32) -> Self {
        Self { state: seed }
    }
    pub fn next_f64(&mut self) -> f64 {
        self.state = self.state.wrapping_mul(1664525).wrapping_add(1013904223);
        (self.state as f64) / 4_294_967_296.0
    }
}

pub fn shuffle(items: &mut Vec<Card>, rng: &mut Rng) {
    let n = items.len();
    for i in (1..n).rev() {
        let j = (rng.next_f64() * (i as f64 + 1.0)) as usize;
        items.swap(i, j);
    }
}

pub fn sort_hand(hand: &mut Vec<Card>) {
    hand.sort_by(|a, b| {
        let sa = SUIT_ORDER[a.suit() as usize].cmp(&SUIT_ORDER[b.suit() as usize]);
        sa.then(a.rank().cmp(&b.rank()))
    });
}

pub fn is_queen_of_spades(c: Card) -> bool {
    c.suit() == SPADES && c.rank() == 12
}
pub fn is_jack_of_diamonds(c: Card) -> bool {
    c.suit() == DIAMONDS && c.rank() == 11
}
pub fn is_two_of_clubs(c: Card) -> bool {
    c.suit() == CLUBS && c.rank() == 2
}
pub fn is_heart(c: Card) -> bool {
    c.suit() == HEARTS
}
pub fn is_penalty(c: Card) -> bool {
    is_heart(c) || is_queen_of_spades(c)
}

pub fn find_card(hand: &[Card], id: u8) -> Option<Card> {
    hand.iter().find(|c| c.id == id).copied()
}

pub fn without_cards(hand: &[Card], ids: &[u8]) -> Vec<Card> {
    hand.iter().filter(|c| !ids.contains(&c.id)).copied().collect()
}
